proto-lens/src/Data/ProtoLens/Encoding/Parser/Internal.hs

```diff
@@ -0,0 +1,40 @@
+-- | Definition of the parsing monad, plus internal
+-- unsafe functions.
+module Data.ProtoLens.Encoding.Parser.Internal
+    ( Parser(..)
+    , ParserResult(..)
+    ) where
+
+import Control.Monad (ap)
+import Control.Monad.Trans.Except
+import Data.Word (Word8)
+import Foreign.Ptr
+
+-- | A monad for parsing an input buffer.
+newtype Parser a = Parser
+    { unParser :: Ptr Word8 -- End position of the input
+               -> Ptr Word8 -- Current position in the input
+               -> ExceptT String IO (ParserResult a)
+    }
+
+data ParserResult a = ParserResult
```

Comment:
How do you feel about renaming this to `ParserSuccess`? I paused for a second to consider why parser result didn't represent a failure case.

Reply:
Makes sense, but I'll do that as a separate PR since it's not really related to this change.

I'll also try benchmarking a change your comment made me realize: we can merge the Either and ParserResult.  Instead of `Either String (ParserResult a)`, have

```
data ParserResult a
   = ParserFailure String
   | ParserSuccess {...}
```